_posts/2025-09-05-anatomy-of-vllm.md

```diff
@@ -0,0 +1,992 @@
+3. The grammar for this request is compiled asynchronously.
+4. During scheduling, if the async compile has completed, the status switches to <code>WAITING</code> and <code>request_id</code> is added to <code>structured_output_request_ids</code>; otherwise it's placed in <code>skipped_waiting_requests</code> to retry on next engine step.
+5. After the scheduling loop (still inside scheduling), if there are FSM requests, the <code>StructuredOutputManager</code> asks the backend to prepare/update <code>_grammar_bitmask</code>.
+6. After the forward pass produces logits, xgr_torch_compile's function expands the bitmask to vocab size (32x expansion ratio because we use 32 bit integers) and masks disallowed logits to –∞.
+7. After sampling the next token, the request's FSM is advanced via <code>accept_tokens</code>. Visually we move to the next state on the FSM diagram.
+
+Step 6 deserves further clarification.
+
+If <code>vocab_size = 32</code>, <code>_grammar_bitmask</code> is a single integer; its binary representation encodes which tokens are allowed ("1") vs disallowed ("0"). For example, "101…001" expands to a length-32 array <code>[1, 0, 1, ..., 0, 0, 1]</code>; positions with 0 get logits set to –∞. For larger vocabularies, multiple 32-bit words are used and expanded/concatenated accordingly. The backend (e.g., <code>xgrammar</code>) is responsible for producing these bit patterns using the current FSM state.
+
+> [!NOTE]
+> Most of the complexity here is hidden in the 3rd party libs like xgrammar.
+
+Here is an even simpler example with vocab_size = 8 and 8-bit integers (for those of you who like my visuals):
+
+<p align="center">
+<picture>
+<img src="/assets/figures/2025-vllm-anatomy/fsm.png" width="100%">
+</picture><br>
+<b>Figure 10</b>: Toy example
+</p>
+
+You can enable this in vLLM by passing in a desired <code>guided_decoding</code> config.
+
+### Speculative Decoding
+
+In autoregressive generation, each new token requires a forward pass of the large LM. This is expensive — every step reloads and applies all model weights just to compute a single token! (assuming batch size == 1, in general it's <code>B</code>)
+
+Speculative decoding [[8]](#ref-8) speeds this up by introducing a smaller draft LM. The draft proposes <code>k</code> tokens cheaply. But we don't ultimately want to sample from the smaller model — it's only there to guess candidate continuations. The large model still decides what's valid.
+
+Here are the steps:
+
+1. <b>Draft</b>: run the small model on the current context and propose <code>k</code> tokens
+2. <b>Verify</b>: run the large model once on context + <code>k</code> draft tokens. This produces probabilities for those <code>k</code> positions plus one extra (so we get <code>k+1</code> candidates)
+3. <b>Accept/reject</b>: going from left to right over the <code>k</code> draft tokens:
+  <ul>
+      <li>If the large model's probability for the draft token ≥ the draft's probability, accept it</li>
```

Comment:
this looks incorrect for rejection sampling. we first sample a uniform distribution, and then compare `p_large(token)/p_draft(token)` with that uniform distribution.

Reply:
that's an implementation detail that doesn't contradict the explanation

it is true that we sample ahead of time vs step by step but i didn't explicitly mention that above